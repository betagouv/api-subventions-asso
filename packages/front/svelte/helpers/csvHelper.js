export const csvToString = array => array.join("\n");

export const arrayToCsv = array => array.join(";");

export const downloadCsv = (content, filename) => {
    const encodedUri = encodeURI("data:text/csv;charset=utf-8," + content);
    const link = document.createElement("a");
    link.setAttribute("href", encodedUri);
    link.setAttribute("download", `${filename}.csv`);
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
};

export const buildCsv = (headers, rows) => {
    if (!Array.isArray(headers) || !Array.isArray(rows)) return undefined;
    return csvToString([arrayToCsv(headers), ...rows.map(arrayToCsv)]);
};
