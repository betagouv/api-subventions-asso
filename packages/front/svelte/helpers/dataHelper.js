export const valueOrHyphen = value => value || "-";
