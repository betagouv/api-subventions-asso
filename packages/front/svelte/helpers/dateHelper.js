import { capitalizeFirstLetter } from "./textHelper";

export const sortByDateAsc = (a, b) => {
    if (!a.date) return 1;
    if (!b.date) return -1;
    return a.date - b.date;
};

export const withTwoDigitYear = date => {
    if (date instanceof Date) {
        const localeDate = date.toLocaleDateString();
        const last2Digit = localeDate.slice(-2);
        const dateWithoutYear = localeDate.slice(0, 6);
        return dateWithoutYear.concat(last2Digit);
    } else return date;
};

export const dateToDDMMYYYY = _date => {
    const date = new Date(_date);
    return `${getDayWithZero(date.getDate())}/${getMonthWithZero(date.getMonth() + 1)}/${date.getFullYear()}`;
};

export const getDayWithZero = day => ("0" + day).slice(-2);

export const getMonthWithZero = day => ("0" + day).slice(-2);

export const isValidDate = date => date instanceof Date && !isNaN(date);

export const getOneYearBeforeDate = date => {
    const prevDate = new Date(date);
    prevDate.setFullYear(date.getFullYear() - 1);
    prevDate.setDate(date.getDate() + 1);
    return prevDate;
};

export const YEAR_START_LOGS = 2021;

export const YEAR_CHOICES = [];
for (let year = YEAR_START_LOGS; year <= new Date().getFullYear(); year++) YEAR_CHOICES.push(year);

export function monthCapitalizedFromId(monthId) {
    const monthStr = new Date(Date.parse(monthId + 1 + " 1, 2022")).toLocaleDateString(`en`, { month: "long" });
    return capitalizeFirstLetter(monthStr);
}
