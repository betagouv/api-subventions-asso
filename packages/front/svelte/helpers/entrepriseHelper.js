export function isAssociation(code_juridique) {
    const LEGAL_CATEGORIES_ACCEPTED = [
        // TODO: créer un package shared pour partager les cat juridique entre le front et le back
        "9210", // Association non déclarée
        "9220", // Association déclarée
        "9221", // Association déclarée d'insertion par l'économique
        "9222", // Association intermédiaire
        "9223", // Groupement d'employeurs
        "9230", // Association déclarée, reconnue d'utilité publique
        "9240", // Congrégation
        "9260", // Association de droit local (Bas-Rhin, Haut-Rhin et Moselle)
        "9300", // Fondation
        "92" // Association loi 1901 ou assimilé
    ];

    return LEGAL_CATEGORIES_ACCEPTED.includes(code_juridique);
}
