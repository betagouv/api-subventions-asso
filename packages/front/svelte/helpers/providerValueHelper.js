// TODO: make this file available to API and FRONT ? => Create a core/shared package ? Put it in DTO package because PV is related to DTO ?

export const flatenProviderValue = providerValueObject => {
    const reduceProviderValues = (acc, prop) => {
        return {
            ...acc,
            [prop]:
                isProviderValues(providerValueObject[prop]) || isProviderValue(providerValueObject[prop])
                    ? getValue(providerValueObject[prop])
                    : flatenProviderValue(providerValueObject[prop])
        };
    };

    if (["number", "string"].includes(typeof providerValueObject) | !providerValueObject) return providerValueObject;

    if (Array.isArray(providerValueObject)) {
        return providerValueObject.map(ob => flatenProviderValue(ob));
    }

    const providerValueObjectProps = Object.keys(providerValueObject);
    return providerValueObjectProps.reduce(reduceProviderValues, {});
};

export const getValue = pv => {
    return getPropFromProviderValue("value")(pv);
};

export const getProvider = pv => {
    return getPropFromProviderValue("provider")(pv);
};

export const getDate = pv => {
    return getPropFromProviderValue("last_update")(pv)?.toString();
};

export const getPropFromProviderValue = value => data => {
    if (!data) return undefined;
    if (Array.isArray(data)) {
        if (data.length === 0) return undefined;
        return data[0][value];
    }
    return data[value];
};

export const isProviderValue = value => {
    return !!value && typeof value == "object" && "value" in value && "provider" in value;
};

export const isProviderValues = values => {
    return Array.isArray(values) && !!values.filter(v => v).length && values.every(v => isProviderValue(v));
};
