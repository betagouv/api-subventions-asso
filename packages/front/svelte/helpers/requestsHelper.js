export function toQueryString(queryObj) {
    return Object.entries(queryObj)
        .reduce((query, [key, value]) => {
            if (value === undefined) return query;
            return query.concat(`${key}=${value}`);
        }, [])
        .join("&");
}
