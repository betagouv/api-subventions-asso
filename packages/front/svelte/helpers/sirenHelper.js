export function siretToSiren(siret) {
    return siret.slice(0, 9);
}
