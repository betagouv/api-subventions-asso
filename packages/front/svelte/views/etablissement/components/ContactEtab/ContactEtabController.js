import { buildCsv, downloadCsv } from "../../../../helpers/csvHelper";
import { formatPhoneNumber, valueOrHyphen } from "../../../../helpers/dataHelper";
import { writable } from "svelte/store";

export default class ContactEtabController {
    constructor(contacts) {
        this._contacts = contacts.map(this.formatContact);
        this._filteredContacts = writable(this._contacts);
        this._roles = this.getContactRoles();
        this._filter = "";
        this.selectedRole = "";
        this._inputName = "";
    }

    getHeaders = () => ["Civilité", "Nom", "Prénom", "Téléphone", "Email", "Rôle"];

    get contacts() {
        return this._filteredContacts;
    }

    get roles() {
        return this._roles;
    }

    get inputName() {
        return this._inputName;
    }

    set inputName(v) {
        this._inputName = v;
        this.filterContacts();
    }

    set filter(f) {
        this._filter = f;
    }

    get filter() {
        return this._filter;
    }

    formatContact = contact => {
        const _contact = { ...contact };
        _contact.telephone = formatPhoneNumber(_contact.telephone);
        for (const property in _contact) {
            _contact[property] = valueOrHyphen(_contact[property]);
        }
        return _contact;
    };

    filterByRole = index => {
        this.selectedRole = this.roles[index];
        this.filterContacts();
    };

    resetContacts = () => {
        this._filteredContacts.update(() => this._contacts);
    };

    filterContacts = () => {
        if (this.selectedRole === "" && this._inputName === "") this.resetContacts();
        else
            this._filteredContacts.update(() =>
                this._contacts.filter(contact => this.containRole(contact) && this.containName(contact))
            );
    };

    containRole = contact => !this.selectedRole || contact.role === this.selectedRole;

    containName = contact =>
        contact.nom.toLowerCase().includes(this._inputName) || contact.prenom.toLowerCase().includes(this._inputName);

    getContactRoles = () =>
        this._contacts.reduce(
            (acc, curr) => {
                if (acc.includes(curr.role)) return acc;
                acc.push(curr.role);
                return acc;
            },
            [""]
        );

    contactToArray = contact => Object.values(contact);

    download = () => {
        downloadCsv(buildCsv(this.getHeaders(), this._contacts.map(this.contactToArray)));
    };
}
