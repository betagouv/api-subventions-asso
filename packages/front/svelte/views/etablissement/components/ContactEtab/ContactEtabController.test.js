import { formatPhoneNumber, valueOrHyphen } from "../../../../helpers/dataHelper";
import ContactEtabController from "./ContactEtabController";

jest.mock("../../../../helpers/dataHelper.js", () => ({
    ...jest.requireActual("../../../../helpers/dataHelper.js"),
    valueOrHyphen: jest.fn(v => v),
    formatPhoneNumber: jest.fn(v => v)
}));

describe("ContactEtabController", () => {
    const CONTACTS = [
        {
            civilite: "Madame",
            nom: "DOE",
            prenom: "Jane",
            telephone: "06 01 02 03 04",
            email: "coordination@foo.bar",
            role: "Coordinatrice"
        },
        {
            nom: "MARTIN",
            prenom: "François",
            telephone: "06 11 22 33 44",
            email: "contact@foo.bar",
            role: "Président"
        },
        {
            nom: "DOE",
            prenom: "Jane",
            telephone: "06 01 02 03 04",
            email: "coordination@foo.bar",
            role: "Coordinatrice"
        },
        {
            nom: "BLACK",
            prenom: "Jeff",
            telephone: "07 11 22 33 57",
            email: "coordination@foo.bar",
            role: "Coordinatrice"
        }
    ];

    let controller;

    beforeEach(() => {
        controller = new ContactEtabController(CONTACTS);
        jest.clearAllMocks();
    });

    describe("_containsRole()", () => {
        it("should return true with role undefined", () => {
            const expected = true;
            const actual = controller._containsRole(CONTACTS[0]);
            expect(actual).toEqual(expected);
        });

        it("should return true with role selected", () => {
            const expected = true;
            controller.selectedRole = "Coordinatrice";
            const actual = controller._containsRole(CONTACTS[0]);
            expect(actual).toEqual(expected);
        });

        it("should return false", () => {
            const expected = false;
            controller.selectedRole = "Président";
            const actual = controller._containsRole(CONTACTS[0]);
            expect(actual).toEqual(expected);
        });
    });

    describe("_getRoles()", () => {
        it("should return an array of roles", () => {
            const expected = ["", "Coordinatrice", "Président"];
            const actual = controller._getRoles();
            expect(actual).toEqual(expected);
        });
    });

    describe("_containsName()", () => {
        it("should return true with empty inputName", () => {
            const expected = true;
            controller._inputName = "";
            const actual = controller._containsName(CONTACTS[0]);
            expect(actual).toEqual(expected);
        });
        it("should return true", () => {
            const expected = true;
            controller._inputName = "doe";
            const actual = controller._containsName(CONTACTS[0]);
            expect(actual).toEqual(expected);
        });
        it("should return false", () => {
            const expected = false;
            controller._inputName = "jeff";
            const actual = controller._containsName(CONTACTS[0]);
            expect(actual).toEqual(expected);
        });
    });

    describe("_filter()", () => {
        it("should filter contacts on name", () => {
            let actual;
            controller._inputName = "doe";
            controller._filter();
            controller.contacts.subscribe(value => (actual = value));
            expect(actual).toMatchSnapshot();
        });

        it("should filter contacts on role", () => {
            let actual;
            controller.selectedRole = "Coordinatrice";
            controller._filter();
            controller.contacts.subscribe(value => (actual = value));
            expect(actual).toMatchSnapshot();
        });
    });

    describe("_format()", () => {
        it("should return format telephone", () => {
            const expected = 1;
            controller._format(CONTACTS[0]);
            const actual = formatPhoneNumber.mock.calls.length;
            expect(actual).toEqual(expected);
        });
        it("should replace empty value with hyphen", () => {
            const expected = 6;
            controller._format(CONTACTS[0]);
            const actual = valueOrHyphen.mock.calls.length;
            expect(actual).toEqual(expected);
        });
    });
});
