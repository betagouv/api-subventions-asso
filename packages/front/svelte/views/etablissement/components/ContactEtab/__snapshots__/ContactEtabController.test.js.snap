// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactEtabController filterContacts should filter contacts on name 1`] = `
[
  {
    "email": "coordination@foo.bar",
    "nom": "DOE",
    "prenom": "Jane",
    "role": "Coordinatrice",
    "telephone": "06 01 02 03 04",
  },
  {
    "email": "coordination@foo.bar",
    "nom": "DOE",
    "prenom": "Jane",
    "role": "Coordinatrice",
    "telephone": "06 01 02 03 04",
  },
]
`;

exports[`ContactEtabController filterContacts should filter contacts on role 1`] = `
[
  {
    "email": "coordination@foo.bar",
    "nom": "DOE",
    "prenom": "Jane",
    "role": "Coordinatrice",
    "telephone": "06 01 02 03 04",
  },
  {
    "email": "coordination@foo.bar",
    "nom": "DOE",
    "prenom": "Jane",
    "role": "Coordinatrice",
    "telephone": "06 01 02 03 04",
  },
  {
    "email": "coordination@foo.bar",
    "nom": "BLACK",
    "prenom": "Jeff",
    "role": "Coordinatrice",
    "telephone": "07 11 22 33 57",
  },
]
`;
